### ISSUE TITLE
List floating clients of (current) tag

### ISSUE THREAD

**Issue by tiosgz** [herbstluftwm/herbstluftwm#831]
It would be nice to have a command to list floating clients (their IDs) of a tag (possibly as a part of `dump` — but how?). This is good e.g. for
* listing all clients of current tag (without calling `attr clients.*.tag` and (if needed) `attr clients.*.floating`), also it could ease listing them in the `cycle_all` order (and this is what I want to use it for)
* restoring *complete* layout of a tag within a session (i.e. when the winids are valid)

**Comment by t-wissmann** [comment]
I'm not sure how to extend the `dump`/`load` command without breaking compatibility. The easiest solution would be to implement a new command
```
list_clients [--floating|--tiling|--all-monitors|--all-tags]
```
But I'm not sure what the output should be. One window id per line? or additional information like the floating state or the window title? (`wmctrl -l` does something very similar, but it can't distinguish between floating and tiled clients).

**Comment by tiosgz** [comment]
> I'm not sure how to extend the `dump`/`load` command without breaking compatibility.

Options (for `dump`, `load` could use autodetection)? Also see below how it could work.
> The easiest solution would be to implement a new command
> 
> ```
> list_clients [--floating|--tiling|--all-monitors|--all-tags]
> ```

Short options `ftAa`? Also, a tag/monitor parameter would be good (empty/missing = focus).
> But I'm not sure what the output should be. One window id per line?

Could be. Sorted by the tag layout (i.e. how they come in `cycle_all`). Other possibility is dump-like `(tag [tiled|floating]:f|t (TILED_DUMP) (floating :N WINID...))` (but this would need to improve a lot) (I'd welcome this option as it's more universal, but it's up to you).
> or additional information like the floating state or the window title?

Only `floating|tiled` if anything. These properties change for every use and are easy to get via `attr` (while the order not).
> (`wmctrl -l` does something very similar, but it can't distinguish between floating and tiled clients).

Unfortunately, this seems to order them by ID and would need a filter so it's unusable for me

**Comment by t-wissmann** [comment]
There also seems to be a need for `list_clients --in-frame`

**Comment by etosan** [comment]
Sorry for not finding this issue sooner, I think this one exactly addresses one of the few remaining pain points I was often hitting with herbstluftwm and that is window collection listing for the purpose of picking and management.

I also agree that proposed dump format of one xid in HEXID form per window per line is sufficient and very scripting friendly. For more detailed output dump is already there.

Such xid list is very easy to split on line boundaries, ids are unambiguous and can then be passed directly to x11 window processing and manipulation tools without too much hassle. Also empty string means empty list of windows, no ambiguous parsing is needed. Getting titles and other properties is easy once we have xid.

In #1249 the aim was to get actionable list of windows from herbst directly, without the need to groom it further through sed and awk or do other expensive walks through object store with `get_attr` - because all these approaches turned out excruciatingly slow (especially when implemented from sh/dash). First there was lag in collecting information from herbst and then it took some time for additional information to be collected by x11 automation tools. With xids provided directly, the whole picker display logic became fast enough for list to appear almost instantaneously in `rofi`. 

In my case the speed could be increased further by moving logic into faster language like lua (which is on my todo list some time, I already have lua extension library in C to communicate with herbst without the `hc` fork/exec overhead directly through x11, but I am not sure if it still works, I haven't tried since winterbreeze came along). 

Anyway this proposal has much better usability and similarly to #1249 is moving the collection selection logic out from script layer into herbst itself, which is the tool best positioned in the desktop stack to do such job quickly. More over it consolidates various collections into single place, which is really, really neat.

In my use cases I needed `--in-frame` to be fastest, but seeing other selectors proposed here made me realize, with this one command, we could ditch quite a lot of other tools and duct tape scripting. We could then get window props quickly with, for example, xprop and order windows around with hc only.

As @tiosgz, I believe that `--in-tag` and `--in-monitor` would be indispensable, as it would allow us to do more with less (but force you to code more as well) :).

Thinking about this got me so very excited about possibilities, that I let my mind roam free, so sorry for next section. 

I believe the best format would something along `list_clients ${selector} [${object_id}] [${modifier}]`. 

Where selector would be one of global ones `--all-monitors|--all-tags` or one of targeted ones `--in-monitor|--in-tag|--in-frame`. Targeted selectors could operate on currently focused entity by default, but it would be very nice if they would also be capable of utilizing standard herbst id semantics for $object_id (at least in case of monitors/tags like -1=left/prev, +1=right/next, number = direct). In case of frames, besides focused, situation gets quite complex really quickly, but imagine if you could do `--list-frame -l|-r|-u|-d|$framepath`. 

Finally without modifier all windows would be returned, but with modifier more precise picks could be easily made. At first I thought `--floating|--tiled|--fullscreen|--urgent` with words for other possible matches, would be fine, but then I realized, that perhaps with `--with-attr something [op something]` it would be more generic better/easier? Even crazier idea: maybe even incorporate parts of matching engine from `hc rule` (not sure if the matcher part from there could be easily reused)? Like  `list_clients --in-frame --match class=urxvt`?

Basically if this command ended up being something between `hc list_frame` and `hc rule` I think herbstluftwm's abilities would end up being pretty phenomenal. I predict with such versatile `list_clients` implementation we could ditch almost every other x automation tool, or use them only in very specific cases. 99% things one would need to do would be doable with herbst only.

**Comment by tiosgz** [comment]
> `--in-tag` and `--in-monitor`

Are you aware that `--in-monitor` is different only by picking the tag that is currently active on the monitor?

---

I must admit that I’m already lost in this issue (maybe because it was opened almost a year ago and I forgot to leave comments in the ‘code’?)

- How should the command look like?
  - Does it even make sense asking this question before the output is clear?
- How should the output look like?
  - I insist on it being in the order HLWM has it in :)
    - That is, if only one tag is considered, in the order they come while doing `cycle_all` from the very first client.
    - With several tags considered (if it ever happens), it would probably be in the order that the tags are in.
  - As far as I understand, the `dump`-like format has been dismissed.
  - One winid per line seems easy and generic.
  - After reading last part of @etosan’s essay, I got another idea, but I’m not sure how much sense it makes and if it is worth it.
    - If the command is called on its own, it just outputs the winids.
    - It can be used in a more complex command, cycled-through with (something like) `foreach`.
      - I’m not sure
        - if the substituted object should be the winids or the `clients.*.` objects
        - how it would be composed into the complex command.
        - Thus this crazy idea should probably be dismissed as well.

**Comment by etosan** [comment]
Phew another long one, sorry. But let me humor you.

In my opinion, let us first remind us what I believe is the real issue here: getting collections of windows quickly (as in fast/performance) and easily (as in conveniently and simply).

Preferably thus, there would be these three major design constraints:
 * fast/performance
 * simplicity (even if apparent only)
 * convenience

These can be evolved further into these basic design rules:
 * command should "appear" "simple", ie. making most common queries should not require us to jump though hoops like it does now (= convenience, simplicity)
 * directly from herbstlutwm. (like no wrapper script running in other process walking the object store = fast/performance, simplicity)
 * immediately actionable (no or extremely minimal collection output parsing, ie without reliance on specialized tooling (ergo without dependecy on sed or awk) = convenience, fast/performance, simplicity)
 * limit "pre-collection" "multipass" hc as much as possible. ( = convenience, fast/performance, simplicity)
 * as generic and versatile as is possible and reasonable. ( = convenience)

So in my eyes this command should appear relatively simple on the surface but be as powerful as it can.

> * As far as I understand, the dump-like format has been dismissed.
> * One winid per line seems easy and generic.

I understand it as such as well, and hope we are on same (and proper) page here. This relates to simplicity and convenience.

> But I'm not sure what the output should be. One window id per line? or additional information like the floating state or the window title? (`wmctrl -l` does something very similar, but it can't distinguish between floating and tiled clients).

I would not complicate the thing with any addtional text in collection output as then it would lose its simple and convenient properties - which are most valuable.

Parsing is PITA, and as both djb and s6 author properly point out, quite dangerous even. It looks simple and easy on the surface, but it is so extremely hard, it led to birth of whole areas of computer science like grammars, finite state machines
and all other crazy stuff. Even herbstclient sidesteps this nicely most of the time, by just packing the argv vector, and sending it to herbstluftwm over x11 wire as is. I was suprised to see git internal dispatcher do same thing: just packing it as "\0" delimited tokens accompanied with argv and sending it over to the next stage. That technique is very powerful and dead simple. So why complicate it?

In this case, instead of "\0", "\n" is acting as token (xid) separator, which is very good compromise. Even the dumbest tools and languages know how to split string on "\n" giving user access to individual collection "records". As soon as I have to
"fish out" both xid and the state from the returned collection record, I already lost everyhting I gained.

In addition, once we have xids, it's very easy to ask herbst anything we want to know about the windows in question through `hc get_attr` interface. As that hc "call" is happening "post-collection", it does not count into: no "pre-collection" hc "multipass" rule,
and frankly, at least since introduction of `get_attr`, asking herbst about it's windows is a non-problem. What is still a major problem is getting the windows collections themselves first.

With regard to getting those collections, there are now two remaining major issues: collection order and collection filtering.

I believe the order should be most hardcoded thing here.

> * I insist on it being in the order HLWM has it in :)
>   * That is, if only one tag is considered, in the order they come while doing cycle_all from the very first client.

I assume this is the same order `focusedFrame->foreachClient()` method uses, and if yes, this was the best order in my `list_frame` attempt.

It maintains most "current" window on top, if remember it right, which is exactly what we need most of the time. I would suggest to support this order only (and perhaps, but only perhaps reverse order - but that can always be achieved with `tac` system command, post-collection, and would simplify already a complex design a lot).

I believe it also avoids complicating the iterator. Order "slightly" breaks anyway, once filtering enters the picture.

> * How should the command look like?
*  I personally believe by now, that it should have dispatcher design similar to git dispatch, i.e. there is initial dispatch, based on first parameter, from which "container" object "class" we want to get windows collection from is obtained.
  * This allows, for example, when reading config (or) scripts to quickly discern one invocation from others based on a first word after `list_clients` only. In my eyes, these are most important object classes, one would want to be able to obtain window collection from: `frame` (this is usecase most important to me), `monitor` (at many points in time I was certainly interested about all the windows on the current monitor, or left monitor, or another monitor), `tag` (while tag and monitor seem to be synonymous they aren't, I think there is subtle difference there)
    * now I am not aware of other container objects of interest but somebody else might be
> * Does it even make sense asking this question before the output is clear?
* I believe it makes, as it limits or expands the basic scope of the command and thus implementation, and having good initial skeleton should be beneficial. However so far I never got too deep into herbst internals like others, because it is quite advanced reading, so I have no idea how various extensions of this command would affect complexity. I think proper implementation of this command will be quite complex.

Now once primary dispatch is done and class is known, for "simplicity" sake, I think default (ie focused) object of that class should be assumed by default, without need to specify it explicitly (this follows behaviour of other `herbstlufwm` commands).
But even if you specify target object explicitly (let's leave out the frames for now, as they are weird, maybe even too complex), like `list_clients --monitor 2`, I believe, we should able to specify only single object at most.
Ie. I think, this one would be confusing: `list_clients --monitor 1 2 3` (like on 4 monitor system). But maybe others may think this is a good idea. I don't know, I would prefer simplicity over convenience here and would defer to multipass hc here.

Now there is also a question, if it would be good, to be able pick collections of all the windows on all defined monitors (`--all-monitors`), and `--all-tags`, which is a superset of `--all-monitors` that contains all the windows in the system (at least as accounted for by herbst).

While these are special cases of previous dispatches, I think it would be good (for convenience): both would allow you, for example, to close all open subtasks related to one giant task on all three (or four) monitors you have, or close all the normal windows in session before logging out, without resorting to session manger to killing them (or xserver going away). Those are just a contrived examples, but I hope they illustrate my point.

> * If the command is called on its own, it just outputs the winids

I understand this idea means, that `list_clients` without any dispatch specifier would be synonymous)with `--all-tags`. While I suggested, we always assume default/focused object after a "container class dispatch", I am not entirely sure that "no dispatch at all" should mean `--all-tags`. I would prefer first, ie dispatch parameter, to be always explicit (ie required), but I know `hc rule` engine consumes even very simple, consequence only rules, like `focus=on`, so that behaviour might make more sense, given the prior art and user expectations. I would still suggest keeping `--all-tags` around as explicit synonym for this usecase (as I would certainly prefer to use it in scripts, to communicate intent to my future self).

One might also say that `--all-monitors` is essentially thin wrapper around `list_clients --monitor 0 1 2` (which I don't think is a good idea in first place) but I think it isn't. 

With the latter, if scripting, you have to go through `get_attr monitors.count` and then `seq+bc`, or `list_monitors`, to get list of monitors first and then do `list_clients` later. In the meanwhile, for whatever reason, the amount of monitors might change during execution (stupid example, I know, we have a locks and other tools to minimize that, but it is illustrative non the less). Maybe this could be chained somehow - but in my eyes it would be easier to just to ask herbst for all windows on all monitors directly.

Same and even more convoluted issues appear when not implementing `--all-tags`/"`list_clients` with no disptach". We don't want to be constructing multiple queries plans,  we just want to able to ask for all the windows that herbst has.

> Are you aware that `--in-monitor` is different only by picking the tag that is currently active on the monitor?

Yes I am, and no I don't think it is supefulous, as `--in-monitor` avoids at least one pre-collection `hc` roundtrip (or in-script state caching) to discover which tag is on which monitor.

It allows you to ask about windows in tag on a any monitor in very effective ways, from which you can build more complex scripted commands very easily (like with `rofi`, `dmenu`, by hand or whatver).

Like for example: asking for windows in current monitor `list_clients --in-monitor`, on specific monitor `list_clients --in-monitor 3`, on previous/next monitor `list_clients --in-monitor -1`, maybe even directionally(?) `list_clients --in-monitor up`. 

With `list_clients --in-monitor` all that can be done without mucking around too much with "pre-collection" hc invocations, "probing first" what monitors we even have and then mapping the tags and monitors together yourself, all that so you could just finally ask `list_clients --in-tag`.

Anyway, with all this said, I think, that the first dispatch part is becoming pretty clear, by now. We can polish it further, but it seems relatively straightforward.

----

Here comes the really hard part: query filtering.

Strictly speaking, filtering is not even neceseary (I am not entirely sure, but once we have initial xid collection, I think we can filter windows by any set of conditions ourselves).

However, where filtering really shines, is insane optimization opportunity it provides. At the cost of internal complexity of course.
If we are able to filter (I remind you optionally) in the moment of collection creation, then we decrease number of times hc has to be called post-collection considerably.

More over we still have an option to filter manually in case of some weird requirements. These factors naturally increases both performance and convenience.

Now how should filtering look like ? I don't have a clue.

But let's say, we all agree `--with-attr` is best candidate that should be implemented, then infinite numbers of permutations `--floating|--tiled|--pseoudtiled` become superfulous, as `--floating` can simply be translated to `list_clients --in-tag --with-attr floating=true`. This form is much more indicative of what is actually happening inside the collection's filtering engine than prebaked `--floating` form. 

Even better it immediately gives us access of all these windows properties:
```
.class
.ewmhnotify
.ewmhrequests
.floating
.fullscreen
.geometry_reported
.instance
.keymask
.keys_inactive
.minimized
.pgid
.pid
.pseudotile
.sizehints_floating
.sizehints_tiling
.tag
.title
.urgent
.visible
.winid
.parent_frame.algorithm
.parent_frame.client_count
.parent_frame.index
.parent_frame.selection
```

That sure, is quite a lot, no matter how you look at it, so it gives users lot of flexibility and freedom, and it also eliminates lot of specialcasing needed for things like `--floating`.

On the other hand ,it requires more generic (and thus more complex) `--with-attr` "executor" engine: there is now a need for operator handling code.

But at least operators `=` and `~` are already implemented for `rule` engine, perhaps even in form, that could be directly usable by `list_clients`.

Unfortunately negations in rule engine are done differently than `!=`, which would be more fitting here.
There are no `</>`, `=</=>` operators. Also shoud there be special `on/off`/`true/false` operators for booleans, or `=true` is better?

Then `rule` engine allows few important filtering selectors that are not exposed as herbtluftwm window attributes: `windowtype` & `windowrole`.

These could be quite useful in certain queries, I am sure, but then it is not `--with-attr` anymore, as there are no such window attributes, really.

Or should we ditch the idea of `--with-attr` and use more intelligent "prebaked" `--with-match/--match` which would know about both attributes and specials like `windowtype`? Such design would mimic `rule` more.

Finally there is a question of how many `--with-attr`/`--with-match/--match` conditions should be allowed: only one or more?
Should this be logical union or should we allow specific logical operators like 'and', 'or' and 'not' in-between?

`rule` engine allows multiple conditions and at least specifc `not` and uses prebaked quivalent of what `--with-match/--match` would be here (but misses few useful operators).

So should `list_clients` matcher follow that design, so that we have proper **POLA** (principle of least surprise), or it can be slightly different?

Price is relatively high, however should all these issues be solved, we get really powerful and very general and universal query engine that can deal with almost any imaginable window collection and that can probably accomodate user's wildest dreams, making even most complex and weird window picks a snap:
 * Want to pick all `urxvts` in a 'max' frame and keep everything else only? no problem, just do...
 * Want to close only windows of specific pid across all the tags? no problem, just do...
 * Want to ...

Due to it's generic nature it would also automatically pick up new window attributes the moment they are added.

> * It can be used in a more complex command, cycled-through with (something like) `foreach`.
>   * I’m not sure
>      *  if the substituted object should be the winids or the `clients.*.` objects
>      *  how it would be composed into the complex command.
>      *  Thus this crazy idea should probably be dismissed as well.

I don't know, I never got around to use `foreach`. I don't use other advanced stuff yet either. For example my chains are very simple and short and not nested. So I am not entirely sure about intricacies of `foreach`.

Also `list_clients` very strongly implies, that we would be getting a list (of xids), so here I would concentrate on maintaining that promise.

However!

I do think that your usecase could still be important to many advanced users (ie those who use chains and `foreach` extensively) and I do believe with proper forethought, maybe, most of the advanced `list_clients` internal machinery (especially the matcher) could be reused, for this use as well. 

Ie. `list_clients` matcher internals would be able to return either string of newline separated xids for `list_clients` OR list 
of `clients.*` objects for `foreach` to wade through. Frontside to this alternative object enabled codepath could maybe be called something like `with_clients`.  It would otherwise maintain same semantcis as `list_clients`, only it would be returning objects instead of xid strings.

Unfortunately, from reading the man page, `foreach` iterator seems to able to just iterate over objects using already builtin collections in their natural order, it does not seem to be able to consume collections coming from somewhere else. But it if is, it would be cool, if it could consume `list_clients/with_clients` machinery output. 

Finally, even if `foreach` couldn't do it, it could still be worthwhile and interesting to have `with_clients` as entirely separate iterator, that would forward objects picked by `list_clients`'s internal matcher to the rest of the chain in the same way as `foreach` does it.

Anyway, nice thing about `list_clients ${class_dispatch} [${object_id}] [${matcher}]` is, I think, that we can start already working 
on the `list_clients ${class_dispatch} [${object_id}]` part, as matcher will require having that code in place anyway.

----

This issue got me playing the whole day with `list_frame` again, as stand in for `list_clients`. I noticed that sometimes it would be quite powerful to be able to do `hc list_clients [--first-only|--last-only] ${dispatch} ...`, but this is probably overgeneralization. What would be your thoughts?

**Comment by tiosgz** [comment]
>> If the command is called on its own, it just outputs the winids
> I understand this idea means, that list_clients without any dispatch specifier would be synonymous)with --all-tags.

I wanted to say that the usual behaviour (`hc list_clients ...`) wouldn’t change, that idea would only affect the cases when the rest of the `hc` call would use the result of `list_clients` (like below)

> (filtering)

I think that *if* filtering should be done, it’d be better when done with the usual HLWM commands syntax – so you only need a way to iterate through the list within one `hc` call.

- you can create as complex filters as you wish
- you can still do so with a single `hc` call
- hlwm-side-wise, it is easier to create than another filtering/matching engine
- on the other hand, I must admit that writing complex hlwm commands is painful

For example, you can do something like this (already works – but includes focused client twice & doesn’t sort them by internal order)

    hc try \ # always give exit status 0, even if last client is not floating
      foreach CLIENT clients. \
        sprintf FLOATING '%c.floating' CLIENT \
          and + compare FLOATING = true \
              + sprintf XID '%c.winid' CLIENT \
                chain , attr XID , echo # 'echo' to separate them (with newlines)

It just needs to be adapted a little, for example like this (assuming that the `!` will tell `foreach` to iterate through a command’s result instead of children in the object tree)

    hc foreach XID !list_clients \
      sprintf FLOATING 'clients.%c.floating' XID \
        and + compare FLOATING = true \
            + chain , attr XID , echo # notice that it is the filter that prints out the winid (or decides what to do with it)

There will, however, be a problem to separate the `list_clients` and `sprintf` commands as soon as `list_clients` gets any arguments. That could be solved by using identifiers like with `chain` etc, in result getting

    hc foreach XID \
      ! list_clients \ # since the '!' is just an identifier, you can use any other string in place of it (as long as it cannot be interpreted as an attribute path)
      ! sprintf FLOATING 'clients.%c.floating' XID \
          and + compare FLOATING = true \
              + chain , attr XID , echo

PS: sorry for making the commands inexecutable by adding comments. I don’t know how to add comments *inside* \*sh commands while keeping them valid.

**Comment by etosan** [comment]
>...
>```
> hc foreach XID !list_clients \
>```

Thank you! I am finally starting to understand how `foreach` works. So this `!` would be extension of `foreach` that would allow it consume arbitrary lists - have you looked at `foreach` internally on how this could be shoehorned inside? I cannot look ATM but I am curious if it consumes lines or objects.

> I think that if advanced filtering should be done, it’d be better when done with the usual HLWM commands syntax – so you only need a way to iterate through the list within one hc call.
> * on the other hand, I must admit that writing complex hlwm commands is painful

This! It's fine for scripts and advaced users I guess, but it gets "little" unpleasant interactively.

> * hlwm-side-wise, it is easier to create than another filtering/matching engine

Certainly, but... I shun advanced commands.  Finally, I would still argue it would be beneficial for users to also have more "normal" filtering engine at their disposal. I certainly, as an user, would want to have it. Let's call it "interactive filtering". Yes, I know, I am very stubborn...

Then looking at your examples I spotted this:

> ```
> ... compare FLOATING = true
> ```

This implies, there is already some fine operator stuff spread inside herbst anyway. This made me realize that maybe all the needed components are already implemented.

So, my naive interpretation is this: `list_clients` will need code to iterate container objects anyway and as I understand it, once that code is in place, the primary, non filtering part of feature is done. 

Should the "interactive filtering" part of this code be implemented, one would need to "lift" args provided by the client, construct some "basic matching function" and evaluate each window being iterated against it and print only xids that match. Problem with that, coding wise is, that this filter code is boring and somewhat "duplication" of what could be already done "just" by chaining already existing commands, right? 

Well, I would like to work something like that, even if in experimental branch and even if it seems wasted effort. It seems it would not require too much intelligence, just gruntwork to connect various parts together to produce all evaluators (at minimum string, number, boolean).

Should such filter be produced, what are the chances of it getting committed into master? 

Or is each superfluous line in herbst seen as maintenance burden later (no code = best code)? 

I know herbst is well maintained, so it might mean going against the project ideals.

> PS: sorry for making the commands inexecutable by adding comments. I don’t know how to add comments inside *sh commands while keeping them valid.

No problem. As far as I know it cannot be done. I would just prepend examples with line `# remove comments when executing in sh` so it's obvious for future readers.

**Comment by tiosgz** [comment]
> have you looked at foreach internally on how this could be shoehorned inside? I cannot look ATM but I am curious if it consumes lines or objects.

Not yet, but I think it iterates through the object tree. The possibility that I suggested could be a separate command – the only reason for using `foreach` is the name.

> It's fine for scripts and advaced users I guess, but it gets "little" unpleasant interactively.

I suggested this because it is a concept that is reusable – there’d be a (new) type internally, e.g. ‘list’, & `foreach` (or some other command) could iterate through it. You can then use this for, let’s say, tags (actually, I’m not sure if there is already something it could be reused for, or if hlwm would have to be extended with new stuff).

> "interactive filtering"

I didn’t like this idea because it seemed too bound to this concrete problem/command. But what if… what if there was a `filter` command instead?

- it would probably work on both the object tree & the new concept of lists
  - the lists would have to be rethought again, since this assumes each item represents an object in the tree as well as the printable value
- it would have a rule-like syntax
  - (maybe it would include all usual operators like `compare`, but in the beginning this isn’t needed)
    - (`and` and `or` also later, if a need arises)
  - instead of having fixed properties to compare, it would compare *only* attributes (i.e. things like window type & role would have to be added to the obj. tree)
    - (again as an idea for the future, testing attributes of child objects)
- yes, all this differs from your idea mostly by using a separate command for filtering, instead of it being a part of `list_clients`

> Should such filter be produced, what are the chances of it getting committed into master?
>
> Or is each superfluous line in herbst seen as maintenance burden later (no code = best code)?
>
> I know herbst is well maintained, so it might mean going against the project ideals.

I’m not the right person to answer this, since I haven’t even seen a continuous piece of its code.

From what I have observed so far, more code doesn’t necessarily mean bad. But OTOH I see the occasional backwards-compatibility fights (including the one at the top of this discussion, which could fortunately be cancelled due to the idea being nonsense), & so I wanted to think of something future-proof.

Anyway, please continue. I may be trying to get something future-proof, but as you can see yourself, it doesn’t always mean that it’s also good.

In the end it will anyway have to pass considerations of @t-wissmann.

**Comment by tiosgz** [comment]
I hope I didn’t scare you away @etosan. If yes, sorry; it wasn’t my intention.

**Comment by etosan** [comment]
No not at all! I just got busy with work. 

Sometimes stuff blows up like crazy and then only thing you want to do is shoot some monsters in the evening and crawl into bed and sleep... Sorry to make you worry @tiosgz!

>I’m not the right person to answer this, since I haven’t even seen a continuous piece of its code. (The less am I an important contributor.)

Me too, this was probably aimed more at @t-wissmann, @The-Compiler and others. But I didn't want to ping them needlessly. Maybe I should have?

I was thinking about it a lot and it basically boils down to two fold issue:
  * quickly gettings xids in various "native collections" (this might be red herring though, as explained in the next part)
  * better/any window filtering engine

These "native collections"  are basically raw groupings we discussed earlier, that only `herbstluftwm` knows about: `--in-frame`, `--in-tag`, potentially `--in-monitor` and others. There is no way for anything else (okay there is: by reconstructing herbst's internal model by arbitrary attr tree walk with external tools, and even that is almost impossible (or very hard) at least in case of frame) to know them. These would be immediately infinitely useful if they could be obtained even without filtering.

But then there is filtering issue. 

Recently I decided to give foreach a try. I wanted get xid of specific single instance window application, but all I had, was the application's pid. I tried this several approaches but in the end ended up doing this monstrosity (please disregard that it's execline script, all you need to know is `pipeline { _left_ } _right_` command is equivalent to ` _left_ | _right_` ):

```shell
#!/usr/bin/execlineb -P

# everything not within { } happens in the same process
getpid
  HC_KP_PID

importas
  -iun
    hc_kp_pid
    HC_KP_PID

background { 
  # here we fork, inside of this block will end up swarming pipleine of 
  # subprocesses that that will run 0.5 sec after main process (outside of this block)
  # is replaced with keepass and their output will be captured into variable
  # and then stored in herbstluftwm object store itself
  if { sleep 0.5 }  # wait while keepass starts
  backtick HC_KP_XID {  # capture stdin into that env variable
    pipeline {
      pipeline {
        pipeline {
          pipeline {
            hc foreach WIN clients. echo WIN  # get clients.$xid list
          } xargs -I@
            execlineb -P -c "backtick PID { hc get_attr @.pid } importas -iun pid PID if { printf \"%s:\" $pid } hc get_attr @.winid" # print PID:XID pairs
        } grep $hc_kp_pid # grep for keepass PID
      } head -n1  # for some reason I always get 2 lines in XID:PID form (why!?)
    } cut -d: -f2   # extract XID of interest
  } importas
      -iun
        hc_kp_xid
        HC_KP_XID
  if { test -n $hc_kp_xid }
    hc set_attr settings.my_keepass_xid $hc_kp_xid # store the xid only if found succesfully
}
```
In the end I added some explanatory comments.

That abomination is roughly equivalent to this shell line:
```shell
hc foreach WIN clients. echo WIN | \
  xargs -I@ execlineb -P -c 'backtick PID { hc get_attr @.pid } importas -iun pid PID if { printf "%s:" $pid } hc get_attr @.winid' | \
    grep $HC_KP_PID | head -n1 | cut -d: -f2
```

I tried to chain `foreach` ~~expansion but I couldn't get my mind around that so if anybody knows how to chain or simplify this I would be immensely grateful~~. 

...

Okay I take that back. While writing this comment, last parts of it clicked into place and I finally got it (example is written in execline because it allows to split argvs with newlines at arbitrary places for readability):
```shell
#!/usr/bin/execlineb -P
hc foreach CLIENT clients. 
  sprintf PID_ATTR "%c.pid" CLIENT 
    sprintf XID_ATTR "%c.winid" CLIENT 
      sprintf PID_XID_PAIR "%s:%s" PID_ATTR XID_ATTR
        echo PID_XID_PAIR
```
Sidenote to self: this chunk is naturally like 1000 times faster than that other pipelined abomination, but dereferencing things with `hc sprintf` is really cumbersome (but kinda cool - it gives off execline vibes).

Getting back to the topic: do I understand it right, that proposed `filter` would be used something like this (am I right?):
```shell
hc filter CLIENT pid="${HC_KP_PID}" echo CLIENT 
# would walk all clients for pid=something and then
# emit something like "clients.0xd34db33f" ?
```

Or even?
```shell
hc foreach CLIENT !filter pid="${HC_KP_PID}" sprinf XID_ATTR "%c.winid" CLIENT sprintf XID "%s" XID_ATTR echo XID
# filter feeds foreach? 
```

That would be cool!

But then, we hit back full circle(!) while it is *relatively* easy to iterate over all windows, which would be roughly equal `list_clients --in-all-tags`, what still remains virtually impossible, again, is how do I do, for example, `filter --in-frame`? 

If you look at it from that angle, I think we are back to square one (but maybe not and something akin to `filter --in-frame` is much better to `list_clients --in-frame`). This is where I constantly end up looping back to and being stuck. Something is missing.

So, to reorient: 
  * We can filter all the windows in a system "relatively" easily even now and we would be able to filter them much more succinctly with `filter` command
  * We *cannot* get list of windows from "native collections" at all currently, and we would be much better off, with any implementation of their listing, than with none. Any implementation, either completely naive as in `list_[clients_in_]frame` / `list_[clients_in_]tag` / `list_[clients_in_]monitor` or more sophisticated as with `list_clients --in-frame|--in-tag|--in-monitor` is better than nothing. So we definitely need this part - at least I certainly do.
  * If listing and filter are to be implemented, than naturally the most useful would be the ability to combine those two features into single resultsets: `listing` + `filter` -> selection of interest. It gives us ability to select any window in any circumstance. This is what we don't (certainly not me) know how to design and implement. To me personally it does not matter how it will end up being possible: `list_clients + filter engine`, `filter + native selection ability` or any other combination.
    * I would prefer for filtering to be less cumbersome than `sprintf` + chaining i.e. `hc rule` engine looks like a sweet design spot for both scripting and most importantly interactive use. But I will accept gladly any other solution that lets me cross correlate and select windows.

I do realize "native collection" idea is little bit wonky. Frames certainly do posses client list, in form of collection, as I have seen them, but lists of clients in tag and in monitor are probably "synthetic" i.e. they appear only in users mind.

----

So thinking about this a lot, maybe the original approach is completely wrong?

This idea came to me this morning: we are probably missing a lot of useful/needed attributes (and perhaps objects).

What if each frame grew some internal identification token, maybe `.frid`? This would not warrant full `frames.0x${frid}` collection, although that could have been really awesome (I also don't know how many attributes could be useful for frame in such collection, if any (maybe just `layout`), nor how this collection should look: like a binary tree or flat list like `clients.`? 

But bear with me please, while full fledged frame collection is maybe overkill, having at least `.frid` in `clients`'s client would allow the `filter` to be given something `filter frid = $FRAME_OF_INTEREST` and voila, now I don't even need `list_clients --in-frame`.

In similar way, if clients grew two new attributes: `.tagid`, which is basically id number of tag, and perhaps: `.monitorid`, `.tagid`'s analogue for monitors, then both  `list_clients --in-tag` and `list_clients --in-monitor` become superfluous, because now I can  relatively comfortably select clients just with `foreach`, or, prefreably, `filter`.  

Keep in mind, that I don't know, if there is possibility for client being known to a system (i.e. to be in `clients.` collection), but not exist within `frame`, `tag` or `monitor`. And if it is possible, I don't know, how to represent such broken links in respective "linking" attribute (`frid`, `tagid`, `monitorid`): maybe by `-1` or `nil`/`null`? Empty string seems to be herbst's norm, but is it?

Sidenotes: poking around `client`  object, I re-discovered there is also a `.tag` attribute, but I would still prefer to have `.tagid` as well, to be able to not rely on user provided names. Since there is `tag` in each `client`, I guess the possibility of window in `clients.` not being bound to some `tag` is thus zero. Now for the `monitor` though, the issue of client not being (currently) on any `monitor` remains, so perhaps the whole idea of single pass filtering by the `monitor` is flawed? Or just empty `.monitorid` is fine deal?

Also, while ids of tags and monitors are somewhat  "fishable" from `tags.` and `monitors.` collections, it would help to have actual `tag.` and `monitor.` objects to have `tag.focus.tagid`/`tag.0.tagid` and `monitors.focus.monitorid`/`monitors.0.monitorid` attribute, same way, like `clients.focus.winid`/`clients.0xbadb3f.winid` have it. This way we would minimize the fishing needed down to a single query and we would also make `tag` and `monitor` object behave more like `client`, which helps consistency. More over, I believe, we would benefit greatly from addition of another new attribute: `monitors.focus.tagid`, that would allow us to go through `tags.$(monitor.tagid).*` directly, than jumping through `tags.by-name.$(monitor.tag).*`. Probably cosmetics, but worthy of pursuit, in name of consistency.  Maybe `monitors.focus.tagid` would be enough to get rid off all the `.monitorid`s in clients?

Anyway, I believe all these modifications would be still relatively minor edits.

Next, the clients themselves would end up having to have three new attributes `.frid` / `.frameid`, `.tagid`, `.monitorid`(questionable but maybe valuable) added. Now, I don't know, how flexible the object tree system is, all I know it used to use HSAttribute object in the code (or something like that), so maybe updating all these atribute in sync with window management events would be problematic (but probably not, as it already tracks a lot of stat,e like `.floating` and it seems highly dynamic)? 

But this part should be somewhat doable too, no?

What now becomes the biggest problem, is, how to represent frames and make them user discoverable, so that we can ask them their `.frid`s, so we are able to use them in filters? Maybe `tag.focus.frid` would suffice? But what about other cases, when I would like to ask about other frames? Hmm it seems like fully fledged frames collection would still be best addition then. And I wanted to avoid it :(!

Finally, `frames` collection design is beyond me. Just pondering about it, both "tree" based and "flat list" + `.focus.` ("traditional" herbst) layout would have a lot to offer. Or maybe have both of them :) ?

Anyway this evolutionary path would make it possible for us, I believe, to completely get rid of `list_clients` and keep only `foreach`/`filter` and gain immense flexibility at the cost of some convenience lost.

So here I imagined some sketches with this "filter only" system (and with extended attributes): 

The opening issue would become something like this:
```shell
hc filter @client tag=!focused floating=on printf "%s\n" @client    
# I just discovered that @ character can be part of matching token, so I use ti to denote "variables". 
# This also contains `printf` command which I completely made up, and is equivalent sprintf in everything else, 
# but instead of storing string it into "variable" it emits it right away.
# "!focused" is just a gimmick it would be cool if filter supported some magic like that

TAGID="$(hc get_attr monitors.focus.tagid)"
hc filter @client tagid="${TAGID}" floating=on printf '%s\n' @client    
# this is same, in somewhat more realistic setup


hc filter @client tagid="$(hc get_attr monitors.focus.tagid)" floating=on printf '%s\n' @client    
# onelined? This is probably something what one would like to write in cli
```

My issue would become something like this:
```
hc  filter @client frid=!focused printf '%s\n' @client    

# or
FRID="$(hc get_attr tags.focus.frid)"
hc  filter @client frid=!focused printf '%s\n' @client    

# or even

hc  filter @client frid="$(hc get_attr tags.focus.frid)" printf '%s\n' @client    

# now with all that - get me all the minimized windows in the frame please:
hc  filter @client frid="$(hc get_attr tags.focus.frid)" minimized=true printf '%s\n' @client 
```

I think this is becoming pretty awesome, while still as flexible as possible. 

Now the biggest problem would then be solving frames and adding (and maintaining linking) of all the needed attributes.

And we definitely need `printf` to avoid "dangling" `sprintf echo` idiom.

----

Regarding X window properties embedded inside `client.` object for the purpose of window type filtering, I have no clue how they work atm, but as far I know X window can add or remove X propeties in realtime as it sees fit, at any moment - so it might be problem when interning them - but `client. ` already has `client.title` which seems to work just fine, so maybe this is non issue.

**Comment by tiosgz** [comment]
> No not at all! I just got busy with work.
>
> Sometimes stuff blows up like crazy and then only thing you want to do is shoot some monsters in the evening and crawl into bed and sleep...

I should have thought of that… hope I didn’t push on you.

---

Sorry, (as usually) I kept writing only the theoretical stuff without examples.

> `list_clients --in-frame|--in-tag|--in-monitor`

This is the part that I want to keep in `list_clients`, since it does what’s applicable only to them (unless I’ve missed something).

`filter` should be able to filter *any* collection (not only windows), so I don’t think that syntax like
> `hc filter CLIENT pid="${HC_KP_PID}" echo CLIENT`

is going to work.

How I imagine it (different possibilities, only one would be the final implementation); here I always ask for all terminals (or `Alacritty`s to be exact) in frame `/@p` (parent of focused *leaf* in unfocused half of the tree) in tag `ichi`:

- `filter SEP COMMAND [ARGS...] SEP RULES...` – should be unambiguous, but isn’t so natural
  - `hc filter . list_clients --tag ichi --frame /@p . class=Alacritty`
  - to allow further actions, maybe it could be even:
    `filter SEP IN_COMMAND [ARGS...] SEP [RULES...] [SEP IDENT OUT_COMMAND [ARGS...]]`
    - `hc filter . list_clients --tag ichi --frame /@p . class=Alacritty . CLIENT printf '%c.title' CLIENT` (yes, I borrowed `printf` from you)
- `filter [RULES...] COMMAND [ARGS...]` – there would have to be something to always distinguish rules from commands
  - `hc filter class=Alacritty list_clients --tag ichi --frame /@p`
- maybe there’s a better option out there

I used `--tag` & `--frame` instead of the `--in-*` variants just because that’s what seems more natural *to me*; in the end it doesn’t matter.

> This would not warrant full `frames.0x${frid}` collection, although that could have been really awesome
> […]
> at least `.frid` in clients

Do you know there’s `tags.*.tiling.` and `clients.*.parent_frame.`, & each frame object has its `index` (i.e. address within its tag)?

Tags & monitors have their numerical `index`; monitors have their `tag`.

**Comment by etosan** [comment]
> filter should be able to filter any collection (not only windows), so I don’t think that syntax like
>>`    hc filter CLIENT pid="${HC_KP_PID}" echo CLIENT`
>is going to work.

Ah! So in your mind the `filter` is generic iterator over any kind of the object and it's attributes, and as such it can filter anything, and yes in that case something must feed it and this example is not gonna work. 

My second part in previous post made me realize that if `filter` would run only over clients/windows, we could could get rid of `list_clients` entirely, but you seem to want to keep it, as you see `filter` as truly generic engine - which is IMHO better too. Then, of course something must feed the `filter` with windows to operate on.

I discovered that right after convenience second thing is relative speed - ability to do herbst stuff in-process seems quite a win. By sprintyfying what I could in my setup, I now increased responsiveness of my scripts considerably (and thus "DE").

From that viewpoint I would argue to make `filter` syntax as powerful as possible and thus ability to fire off embedded chain right away would be very valuable.

Both `rule` and `execline` work by "recognizing" the proper words implicitly, `rule` has finite set of words that my appear in it, it seems, while in execline, it's just argv that is being consumed by commands run (similar to how `foreach` operates, it seems). But I guess that is too much uncertainty for `filter`. 

At first I wanted to propose: why not reuse `rule` concept and use `--` as major part section separator? 

But then I agree that using user provided separator would be much more flexible, thus better. So I would insist  on keeping "humane" ordering with separator then. 

I would only rephrase the syntax bit: rename IN_COMMAND to LIST_GENERATOR as that is actually the widget which *generates* the list iterated by `filter`:
`filter SEP LIST_GENERATOR [ARGS...] SEP [RULES...] [SEP IDENT OUT_COMMAND [ARGS...]]

Now regarding `list_clients` I presume it is easy to have it output objects or xids depending on the invocation context?
As in raw `hc list_clients` emits xid, while inside `filter` : `hc filter -- list_clients -- ...` it emits objects?

Then I think this is is essentially becoming quite well designed.

> I used `--tag` & `--frame` instead of the `--in-*` variants just because that’s what seems more natural to me; in the end it doesn’t matter.

As long as `--tag` and `--frame` without any parameter selects focused entities without explicit need to specify them (which is herbst default anyway), it doesn't matter to me either ( I just thought `--in-*` form was more readable). 

I would still welcome if I could do `--monitor MONITOR_DELTA` without jumping through `--tag` in between, especially with "spatial deltas" like `-l`, `-r`, `-u`. `-d`, and I would be immensely glad if frame index, at least in this case supported same "spatial deltas", as monitor one (tags don't really have spatial awareness of their own) but I digress, it's just syntactic sugar. Very nice, but not that essential.

>>This would not warrant full frames.0x${frid} collection, although that could have been really awesome
>>[…]
>>at least .frid in clients
>
>Do you know there’s tags.*.tiling. and clients.*.parent_frame., & each frame object has its index (i.e. address within its tag)?
>
>Tags & monitors have their numerical index; monitors have their tag.

Hmm almost there but not! 

Okay after your pointer I realize that `tag.index` and `monitor.index` are those `.tagid` and `.monitorid` I proposed. Great. But this still does not solve what I wanted to point out.

Also regarding frame index, it seems this is the frame index as described by manual: so on single frame max `layout` it's empty i.e `root` and as such cannot be used as frame identity:
```shell
$hc get_attr tags.focus.tiling.focused_frame.index
$
```
Tagdump:
```
: hc dump 
(clients max:0 0x2009a6) <= urxvtd window
```
Let me explain: I tried through years several ways to calculate intersections (which turns out is second part of my previous post as well), and I think without additional attributes in client this is still impossible: it should be perfectly possible to calculate intersection of all windows/clients against a tag, or a monitor, or a frame - or anything else basically, but it isn't. I believe that `list_clients` is basically in-herbst implementation of this feature.

While I can get tag id or monitor id readily, which really is their `.index`, as you correctly pointed out (so maybe the `client` attrs should be named `.tagindex` and `.monitorindex` instead?) I can never use this information directly against the clients themselves: 
```shell
$ hc complete 1 get_attr clients.focus. 
...
clients.focus.tag # <- this is not a tag object, but a string (which is fine but maybe should be tagname?)
#clients.focus.tagindex # <- could easily be here, there is no reason why not
#clients.focus.monitorindex # <- same here, rereading, manual seems official value should -1 for windows not visible on any monitor
...
clients.focus.parent_frame. # <- this is nested, but I still don't know which frame this is
# okay by dereferencing deeper parent_frame object I can still get a lot of information about the frame but not most importation one: it's identity 
$ hc complete 1 get_attr clients.focus.parent_frame.
clients.focus.parent_frame.algorithm # layout
clients.focus.parent_frame.client_count # number of clients in frame
clients.focus.parent_frame.index # <- detailed layout index, sadly not an identity
clients.focus.parent_frame.selection # <- window order within frame
#clients.focus.parent_frame.frid # <-maybe identity could be here?
```

So with `filter`, and maybe even now with `foreach` and chains and some very heavy thinking - I should be able to take `tag.index` 
and easily pick windows bound to that `tag.index`. I can not actually do it, Well I can, but not in that straightforward way,  now that I finally learned to use `sprintf`, I can somewhat do it, when going through tag name (I switched to '^' which is hopefully free for now as @ can be used in frame.index it seems and other character are taken by shell):

```shell
$ hc foreach ^clients clients. sprintf ^xid_attr "%c.winid" ^clients sprintf ^tag_attr "%c.tag" ^clients sprintf ^xid_tag_map "%s:%s" ^xid_attr ^tag_attr echo ^xid_tag_map | grep $(hc get_attr tags.focus.name) | cut -f1 -d:
0xzzzzz
0xyyyyy
...
```

I would still argue index would be much more robust, as it does not allow arbitrary user characters:

```shell
 $ hc rename  sometag  bad:stuff:can:happen:if:i:use:wrong:xid:tagname:separator:order:with:cut:above
```

But okay. So intersection of clients and tags is somewhat solved. Monitor could be adapted from there too, but would not be doable single pass, like this :( (maybe worthwhile trade-off). 

So now, how can one do this with frames?

I think I should at least be able to singlepass client <-> frame same way like tag above. Lets pretend frames have grown an actual `frid`:
```shell
$ hc foreach ^clients clients. sprintf ^xid_attr "%c.winid" ^clients sprintf ^frid_attr "%c.parent_frame.frid" ^clients sprintf ^xid_frid_map "%s:%s" ^xid_attr ^frid_attr echo ^xid_frid_map | grep $(hc get_attr tags.focus.tiling.focused_frame.frid) | cut -f1 -d:
```

When looping with this back to `filter` question now is, will it allow dereferencing nested attributes like `sprintf` can with '%s' ?
```shell
hc filter -- clients. -- parent_frame.frid=0xzzzzz -- ^client  sprintf ^xid_attr "%c.winid" ^client printf "%s\n" ^xid_attr
```
I am specifically talking about `parent_frame.frid=0xzzzzz` comparison, because if not, then `.frid` should definitely moved "up" into `client`.

But as you pointed out, if deeper dereferencing would be possible, I would be entirely content with `parent_frame|focused_frame.frid`. Thus you are right, `frames.` collection would be superfluous.

Regarding `.frid` assignments/generation - I think nothing complicated should be used, frames are very dynamic entities in herbst so I believe simple uint32_t counter that always grows would be good enough. If I understand it right, this would allow 4 billions unique `.frid`s per desktop session, without wraparound, and when it finally wraps, lower "numbered" frames should not exist. If this seems too wonky I believe uin64_t should be able to express session frames for several ... I don't know how many ... years. Wrong idea?

**Comment by tiosgz** [comment]
> Now regarding `list_clients` I presume it is easy to have it output objects or xids depending on the invocation context?
> As in raw `hc list_clients` emits xid, while inside `filter` : `hc filter -- list_clients -- ...` it emits objects?

I hope so.

> ```
> hc foreach ^clients clients. sprintf ^xid_attr "%c.winid" ^clients sprintf ^tag_attr "%c.tag" ^clients sprintf ^xid_tag_map "%s:%s" ^xid_attr ^tag_attr echo ^xid_tag_map | grep $(hc get_attr tags.focus.name) | cut -f1 -d:
> ```

```
hc substitute ^tag_name tags.focus.name \
   foreach ^clients clients. \
     and + sprintf ^client_tag '%c.tag' ^clients \
             compare ^client_tag = ^tag_name \
         + sprintf ^xid '%c.winid' ^clients \
             chain , get_attr ^xid , echo | sort | uniq
```

(just saying :)) The `sort | uniq` is to remove one appearance of focused client (can be avoided – see last snippet).

Similarly,

> ```
> hc foreach ^clients clients. sprintf ^xid_attr "%c.winid" ^clients sprintf ^frid_attr "%c.parent_frame.frid" ^clients sprintf ^xid_frid_map "%s:%s" ^xid_attr ^frid_attr echo ^xid_frid_map | grep $(hc get_attr tags.focus.tiling.focused_frame.frid) | cut -f1 -d:
> ```

```
hc substitute ^fr_idx tags.focus.tiling.focused_frame.index \
   substitute ^tag_name tags.focus.name \
     foreach ^client clients. \
       and + sprintf ^cl_tag '%c.tag' ^client \
               compare ^cl_tag = ^tag_name \
           + sprintf ^cl_frame '%c.parent_frame.index' ^client \
               compare ^cl_frame = ^fr_idx \
           + sprintf ^xid '%c.winid' ^client \
               chain , get_attr ^xid , echo | sort | uniq
```
Or, pure `hc` (lemme get crazy)
```
hc substitute ^fr_idx tags.focus.tiling.focused_frame.index \
   substitute ^tag_name tags.focus.name \
   mktemp string ^foc_wid \
   chain xxx substitute ^xid clients.focus.winid \
               set_attr ^foc_wid ^xid \
     xxx foreach ^client clients. \
       or / and + sprintf ^xid '%c.winid' ^client \
                  sprintf ^fid '%s' ^foc_wid \
                    compare ^xid = ^fid \
                + set_attr ^foc_wid '' \
          / and + sprintf ^cl_tag '%c.tag' ^client \
                    compare ^cl_tag = ^tag_name \
                + sprintf ^cl_frame '%c.parent_frame.index' ^client \
                    compare ^cl_frame = ^fr_idx \
                + sprintf ^xid '%c.winid' ^client \
                    chain , get_attr ^xid , echo
```

**Comment by t-wissmann** [comment]
I'm really loosing track because it is too much text for me to read. In general I can say that I don't want to implement another command system within hlwm, so if you want to do anything more advanced with the window list, then this should be done by some `foreach_client` command. But for getting things started simple, I plan to implement `list_clients` (hopefully) soon.

**Comment by tiosgz** [comment]
If I’m not mistaken, it can be ‘summed up’ (even the summary turned out to be long, sorry) as:

- `list_clients [--in-tag|--in-frame|--in-monitor [ID]]` (I’m not really sure if the ID makes sense to be optional)
  - `hc list_clients --in-tag ichi --in-frame /@p`
- `filter SEP LIST_GENERATOR [ARGS...] SEP [RULES...] [SEP IDENT OUT_COMMAND [ARGS...]]`
  - `SEP` – the usual hlwm separator (as used in `chain`, `and`…)
  - `LIST_GENERATOR` – as of now only `list_clients`
  - `RULES` – something like `tag=ichi`; each rule checks value of the attribute (`tag`) against the passed value (`ichi`)
  - `IDENT OUT_COMMAND` (optional) then acts like `foreach`, iterating the results
  - `hc filter . list_clients --tag ichi --frame /@p . class=Alacritty . CLIENT printf '%c.title' CLIENT`
- one new thing this requires is some new return type that
  - outputs something (for `list_clients` it’s winIDs) when the command isn’t part of a bigger structure (as in the `list_clients` example)
  - allows being iterated through (always giving an object in the object tree) when it’s combined (as in the other example)
  - → you can `filter` a list of clients, but it’s easy to extend to e.g. tags in the future
- the point of the `filter` command is to avoid zillion-line commands when you want to find objects with one or two attributes of a given value

Does it seem too complex?

**Comment by t-wissmann** [comment]
I'm not sure whether you have noticed it, but the `list_clients` command is now implemented with most of the features you've requested:
```
list_clients [--tag=TAG|--monitor=MONITOR]
             [--frame=FRAME_PATH|--floating|--tiling]
             [--title]
```
(See the man page for more details).

I'm still hesitating to introduce another filtering language just for this; I'd really prefer to have it linked to the existing command system.

**Comment by tiosgz** [comment]
> I'm still hesitating to introduce another filtering language just for this; I'd really prefer to have it linked to the existing command system.

I see the point, as it would introduce a lot of new abstract things. But it doesn’t seem *entirely* unrelated to the current command system (it is similar to `rule`, although that one is much simpler). Well, your WM, your decision (to be honest, I would also be cautious and try to go the path of least resistance). Scripts can do that as well, but they don’t completely satisfy @etosan’s goals of speed and command simplicity.

**Comment by tiosgz** [comment]
With more than a year’s distance, all this seems pretty over-engineered to me. The originally requested feature has long been present in HLWM, therefore i see no reason for not closing this. If someone sees the need for anything discussed later, i suggest them opening a new issue.

**Comment by t-wissmann** [comment]
Thanks! Also, writing custom loops should be easier now because `foreach` now has options that make it easier to write custom loops:

  * `--filter-name=REGEX` consider children whose name match the specified REGEX.
  * `--unique` do not print duplicates (some objects can be reached via multiple paths, such as clients.focus)
  * `--recursive` print OBJECT and all its children of arbitrary depth in breadth-first search order. This implicitly activates --unique.

and the soon to be added `sprintf` format `%{}` (PR https://github.com/herbstluftwm/herbstluftwm/pull/1481) will make the loop body much shorter and simpler.
